File: flutter_embed_unity_2022_3_ios/ios/Classes/View/UnityViewStack.swift
//
//  UnityViewStack.swift
//  flutter_embed_unity_ios
//
//  Created by James Allen on 28/08/2023.
//

import Foundation


// This class is responsible for making sure that Unity is only ever attached to the
// topmost view in the stack. Unity cannot be attached to more than one view, so there
// should only ever be one EmbedUnity widget on a Flutter screen - however we still
// need to account for the fact that when pushing a new Flutter route / screen onto
// the Navigator stack, both screens are still alive and so we can end up with more
// than one PlatformView at the same time (but only the top one will be visible)
// See https://developer.apple.com/documentation/uikit/uiviewcontroller
// See https://developer.apple.com/documentation/uikit/uiview
// See https://developer.apple.com/documentation/swift/using-key-value-observing-in-swift
class UnityViewStack: NSObject {
    
    // This could possibly be implemented as a Queue / Stack collection, but it may
    // be possible that a view which isn't the topmost one gets disposed (eg during
    // a Navigator.of(contect).pushAndRemoveUntil ?) so safest just to use a list
    private var viewStack = [UnityViewController]()
    
    func pushView(_ viewController: UnityViewController) {
        // Unity can only be attached to one view at a time. Therefore, check
        // if there are any other active views, and detatch Unity from them first
        viewStack.forEach { existingViewController in
            existingViewController.detachUnity()
        }
        
        // attach Unity to the new view
        let unityPlayerSingleton = UnityPlayerSingleton.getInstance()
        viewController.attachUnity(unityPlayerSingleton)
        
        // Add view to the stack
        viewStack.append(viewController)
        NSLog("UnityViewStack: pushed Unity view \(viewController.viewId) onto stack")
        
        
        // ---------------------------------------------
        // TODO: find a simpler way to achieve the below
        // The following isn't ideal. The reason is that I couldn't find an
        // obvious way to reliably detect when the viewController is destroyed.
        // FlutterPlatformView doesn't provide a `dispose` method
        // to override (like the Android equivalent does), classes in Swift
        // don't seem to provide a dispose or dealloc method to override, and
        // I can't find anything useful in UIViewController or UIView to detect
        // being destroyed. There is probably some easy way to do this but I just
        // can't figure it out yet.
        // So instead this is a bit of a workaround using viewDidDisappear
        // on UIViewController, which is called if the view is destroyed, OR
        // also if the view is simply being obscured by another view ontop (eg if a
        // Flutter PageRoute is pushed onto the stack). Therefore the below logic
        // needs to handle both cases:
        
        // If it disappears, it MAY have been destroyed, so remove from stack
        viewController.viewDidDisappear = { viewId in
            NSLog("UnityViewStack: Unity view \(viewController.viewId) disappeared, removing from stack")
            self.popView(viewController)
        }
        
        // However it may reappear if it wasn't destroyed (eg it was obscured underneath
        // another screen, and now has reappeared), in which case push it back onto the stack
        viewController.viewDidAppear = { viewId in
            if !self.viewStack.contains(where: {$0.viewId == viewId}) {
                NSLog("UnityViewStack: View \(viewId) has reappeared, pushing back onto stack")
                self.pushView(viewController)
            }
        }
        // ---------------------------------------------
        
        
        // Resume unity
        unityPlayerSingleton.pause(false)
    }

    private func popView(_ viewController: UnityViewController) {
        // Detatch Unity from the view
        viewController.detachUnity()
        // Remove from the stack
        viewStack = viewStack.filter { $0 != viewController }
        NSLog("Detached Unity from popped view")

        let unityPlayerSingleton = UnityPlayerSingleton.getInstance()
        
        if(!viewStack.isEmpty) {
            // If there are any remaining views in the stack, attach Unity to the last view to be
            // added to the stack
            viewStack.last?.attachUnity(unityPlayerSingleton)
            NSLog("Reattached Unity to existing view")
            // I don't know why, but when Unity is reattached to an existing view
            // we need to pause AND resume (even though Unity was never paused?):
            unityPlayerSingleton.pause(true)
            unityPlayerSingleton.pause(false)
        }
        else {
            // No more Unity views, so pause
            NSLog("No more EmbedUnity widgets in stack, pausing Unity")
            unityPlayerSingleton.pause(true)
        }
    }
}

File: flutter_embed_unity_6000_android/android/src/main/kotlin/com/learntoflutter/flutter_embed_unity_android/platformView/UnityViewStack.kt
package com.learntoflutter.flutter_embed_unity_android.platformView

import android.view.WindowManager
import com.learntoflutter.flutter_embed_unity_android.constants.FlutterEmbedConstants.Companion.logTag
import com.learntoflutter.flutter_embed_unity_android.unity.UnityPlayerSingleton
import io.flutter.Log
import io.flutter.util.ViewUtils

/**
 * This class is responsible for making sure that Unity is only ever attached to the
 * topmost view in the stack. Unity cannot be attached to more than one view, so there
 * should only ever be one EmbedUnity widget on a Flutter screen - however we still
 * need to account for the fact that when pushing a new Flutter route / screen onto
 * the Navigator stack, both screens are still alive and so we can end up with more
 * than one PlatformView at the same time (but only the top one will be visible)
 */
class UnityViewStack {
    // This could possibly be implemented as a Queue / Stack collection, but it may
    // be possible that a view which isn't the topmost one gets disposed (eg during
    // a Navigator.of(contect).pushAndRemoveUntil ?) so safest just to use a list
    private val viewStack = ArrayList<IUnityViewStackable>()

    fun pushView(view: IUnityViewStackable) {
        // Unity can only be attached to one view at a time. Therefore, check
        // if there are any other active views, and detatch Unity from them first
        for(existingView in viewStack) {
            existingView.detachUnity()
        }

        // Then attach Unity to the new view
        val unityPlayerSingleton = UnityPlayerSingleton.getOrCreateInstance()
        if(unityPlayerSingleton != null) {
            view.attachUnity(unityPlayerSingleton)

            // It's important to call windowFocusChanged, otherwise unity will not start
            // (not sure why - UnityPlayer is undocumented)
            unityPlayerSingleton.windowFocusChanged(unityPlayerSingleton.getFrameLayout().requestFocus())
            // I don't know why, but when Unity is detached from an existing view and
            // added to a new view, we need to pause AND resume instead of just resume:
            unityPlayerSingleton.pause()
            unityPlayerSingleton.resume()

            // Unity on Android always forces hiding the status bar.
            // See https://forum.unity.com/threads/status-bar-always-hidden-on-android.362779
            // See https://github.com/Over17/UnityShowAndroidStatusBar
            // This is a workaround to show the status bar:
            // TODO: FLAG_FULLSCREEN is deprecated, what to replace it with?
            // unityEngineSingleton.windowInsetsController?.show(WindowInsets.Type.statusBars()) doesn't seem to work...
            ViewUtils.getActivity(unityPlayerSingleton.context)?.window?.clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)

            Log.i(logTag, "Attached Unity to new view")
        }
        else {
            Log.e(logTag, "Unity hasn't been attached to the view because UnityPlayer " +
                    "couldn't be created")
        }

        // Register for onDispose event so we can pop the stack when this view is disposed
        view.onDispose = {
            popView(view)
        }

        // Add view to the stack
        viewStack.add(view)
    }

    private fun popView(view: IUnityViewStackable) {
        // Detatch Unity from the view
        view.detachUnity()

        // Remove from the stack
        viewStack.remove(view)

        if(viewStack.isNotEmpty()) {
            // If there are any remaining views in the stack, attach Unity to the last view to be
            // added to the stack
            val unityPlayerSingleton = UnityPlayerSingleton.getInstance()
            if(unityPlayerSingleton != null) {
                viewStack.last().attachUnity(unityPlayerSingleton)
                Log.i(logTag, "Reattached Unity to existing view")
                // I don't know why, but when Unity is reattached to an existing view
                // we need to pause AND resume (even though Unity was never paused?):
//                unityPlayerSingleton.windowFocusChanged(unityPlayerSingleton.getFrameLayout().requestFocus())
                unityPlayerSingleton.pause()
                unityPlayerSingleton.resume()
            }
            else {
                Log.e(logTag, "Unity hasn't been reattached to the last view in the " +
                        "stack because UnityPlayer was null")
            }
        }
        else {
            // No more Unity views, so pause
            Log.i(logTag, "No more EmbedUnity views in stack, pausing Unity")
            UnityPlayerSingleton.getInstance()?.pause()
            // DO NOT call unityPlayerCustom.destroy(). UnityPlayer will also kill the process it is
            // running in, because it was designed to be run within it's own activity launched in it's
            // own process. We can't make FlutterActivity launch in it's own process, because it's the
            // main (and usually the only) activity.
        }
    }
}
